python/morpheus/morpheus/stages/general/router_stage.py

```diff
@@ -0,0 +1,123 @@
+                 key_fn: typing.Callable[[ControlMessage], str],
+                 is_runnable: bool = False) -> None:
+        super().__init__(c)
+
+        self._keys = keys
+        self._key_fn = key_fn
+        self._is_runnable = is_runnable
+
+        self._router: mrc.core.segment.SegmentObject | None = None
+
+        self._create_ports(1, len(keys))
+
+    @property
+    def name(self) -> str:
+        return "router"
+
+    def supports_cpp_node(self):
+        return True
+
+    def _pre_compute_schema(self, schema: _pipeline.StageSchema):
+        # Pre-flight check to verify that the input type is one of the accepted types
+        super()._pre_compute_schema(schema)
+        input_type = schema.input_type
+        if (not typing_utils.issubtype(input_type, ControlMessage)):
+            raise RuntimeError((f"The {self.name} stage cannot handle input of {input_type}. "
+                                f"Accepted input types: {(ControlMessage,)}"))
+
+    def compute_schema(self, schema: _pipeline.StageSchema):
+
+        # Get the input type
+        input_type = schema.input_type
+
+        for port_idx in range(len(self._keys)):
+            schema.output_schemas[port_idx].set_type(input_type)
+
+    def _build(self, builder: mrc.Builder, input_nodes: list[mrc.SegmentObject]) -> list[mrc.SegmentObject]:
+
+        def _key_fn_wrapper(msg) -> str:
+            return self._key_fn(msg)
+
+        assert len(input_nodes) == 1, "Router stage should have exactly one input node"
+
+        if (self._build_cpp_node()):
+            import morpheus._lib.stages as _stages
+
+            if (self._is_runnable):
+                self._router = _stages.RouterControlMessageRunnableStage(builder,
+                                                                         self.unique_name,
+                                                                         router_keys=self._keys,
+                                                                         key_fn=_key_fn_wrapper)
+            else:
+                self._router = _stages.RouterControlMessageComponentStage(builder,
+                                                                          self.unique_name,
+                                                                          router_keys=self._keys,
+                                                                          key_fn=_key_fn_wrapper)
+        else:
+            from mrc.core.node import Router
+            from mrc.core.node import RouterComponent
+
+            if (self._is_runnable):
+                self._router = Router(builder, self.unique_name, router_keys=self._keys, key_fn=_key_fn_wrapper)
+            else:
+                self._router = RouterComponent(builder,
+                                               self.unique_name,
+                                               router_keys=self._keys,
+                                               key_fn=_key_fn_wrapper)
```

Comment:
Why are the C++ stage impls needed if the base mrc nodes could be used?

Reply:
~~Because the base nodes operate on python objects. So you have to do a cast/uncast for each message. Just compiling for the type is easy and didnt require much effort.~~

I just typed that all out and then realized that we still have to do the cast anyways and grab the GIL for the `key_fn` so this isnt needed anymore. 😔 . I'll remove.

Reply:
Removed the C++ implementations